File: frontend/resources/run_test.py
import io
import os
import subprocess
import sys

# Change standard output encoding to UTF-8
sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')

def run_command(command, log_file):
    with subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.STDOUT) as process:
        with open(log_file, "ab") as f:  # Open in binary append mode
            for line in iter(lambda: process.stdout.readline(), b''):
                print(line.decode('utf-8', errors='replace'), end='')  # Decode for printing
                f.write(line)  # Write raw bytes to file

            if process.wait() != 0:
                print(f"Command '{command}' failed with error")


def main(block_dir, block_key):
    container_dir = '/app'  # The directory inside the container where you want to mount
    image_name = container_name = block_key
    
    commands = [
        ["docker", "rm", container_name],
        ["docker", "rmi", image_name],
        ["docker", "build", "-t", image_name, block_dir],
        ["docker", "run", "--name", container_name, "-v", f"{block_dir}:{container_dir}", image_name, "python", "-B", "-c", "from computations import test; test()"],
    ]
    
    log_file = os.path.join(block_dir, 'logs.txt')
    with open(log_file, 'wb'):
        pass

    for command in commands:
        print(f'Executing: {command}')
        run_command(command, log_file)

if __name__ == "__main__":
    block_dir = sys.argv[1]
    block_key = sys.argv[2]
    main(block_dir, block_key)

File: frontend/resources/runTest.mjs
import { spawn } from "child_process";
import fs from "fs";
import path from "path";

function runCommand(command, args, logFile) {
  return new Promise((resolve, reject) => {
    const process = spawn(command, args);

    const logStream = fs.createWriteStream(logFile, { flags: "a" });

    process.stdout.on("data", (data) => {
      const output = data.toString("utf-8");
      process.stdout.write(output);
      logStream.write(data);
    });

    process.stderr.on("data", (data) => {
      const output = data.toString("utf-8");
      process.stderr.write(output);
      logStream.write(data);
    });

    process.on("close", (code) => {
      logStream.end();
      if (code !== 0) {
        reject(
          new Error(
            `Command '${command} ${args.join(" ")}' failed with error code ${code}`,
          ),
        );
      } else {
        resolve();
      }
    });
  });
}

export async function runTestContainer(blockDir, blockKey) {
  const containerDir = "/app"; // The directory inside the container where you want to mount
  const imageName = blockKey;
  const containerName = blockKey;

  const commands = [
    ["docker", ["rm", containerName]],
    ["docker", ["rmi", imageName]],
    ["docker", ["build", "-t", imageName, blockDir]],
    [
      "docker",
      [
        "run",
        "--name",
        containerName,
        "-v",
        `${blockDir}:${containerDir}`,
        imageName,
        "python",
        "-B",
        "-c",
        "from computations import test; test()",
      ],
    ],
  ];

  const logFile = path.join(blockDir, "logs.txt");
  fs.writeFileSync(logFile, ""); // Clear the log file

  for (const [command, args] of commands) {
    console.log(`Executing: ${command} ${args.join(" ")}`);
    try {
      await runCommand(command, args, logFile);
    } catch (error) {
      console.log(`Command '${command}' failed with error`);
    }
  }
}
